Deploy subgraph (#788)
graphprotocol/docs: website/pages/en/deploying/deploying-a-subgraph-to-studio.mdx

---
title: Deploy Using Subgraph Studio
---

> Learn how to deploy your subgraph in Subgraph Studio.

Reviewer: ```suggestion
Learn how to deploy your subgraph to Subgraph Studio.
```

> Note: When you deploy a subgraph, you push it to Subgraph Studio, where you'll be able to test it. It's important to remember that deploying is not the same as publishing. When you publish a subgraph, you're publishing it on-chain.

## Subgraph Studio Overview

In [Subgraph Studio](https://thegraph.com/studio/), you can do the following:

- Create and manage your API keys for specific subgraphs
- Create your subgraph through the Studio UI
- Deploy your subgraph using the The Graph CLI
- Integrate your subgraph in staging using the query URL
- Publish your subgraph with the Studio UI

> - Test your subgraph in the playground environment

Reviewer: I would put this one before the "integrate your subgraph in staging using the development query URL".

> - Restrict your API keys to specific domains and allow only certain Indexers to query with them

Reviewer: I would put this one after the "create and manage your API keys" one.

- View a list of subgraphs you've created

> - Manage, view details, and visualize the status of a specific subgraph

Reviewer: I would put these two first.

---
title: Deploy Using Subgraph Studio
---

Learn how to deploy your subgraph in Subgraph Studio.

> > Note: When you deploy a subgraph, you push it to Subgraph Studio, where you'll be able to stage it. It's important to remember that deploying is not the same as publishing. When you publish a subgraph, you're publishing it on-chain.

Reviewer: Is "stage" the term we want to use? I feel like "test" is more accessible.

Author: -This change was in response to feedback from devs

## Subgraph Studio Overview

In [Subgraph Studio](https://thegraph.com/studio/), you can do the following:

- Create and manage your API keys for specific subgraphs

> - Create your subgraph through Studio UI

Reviewer: ```suggestion
- Create your subgraph through the Studio UI
```

- Deploy your subgraph using the CLI
- Publish your subgraph with the Studio UI

> - Stage your subgraph in the playground environment

Reviewer: ```suggestion
- Test your subgraph in the playground environment
```

> - Integrate your subgraph in staging using the query URL

Reviewer: ```suggestion
- Integrate your subgraph in staging using the development query URL
```

Also, I think this bullet point (and potentially the previous one as well) should come before "Publish your subgraph with the Studio UI", because using the development query URL to test a subgraph in a staging environment is usually done before publishing it to the network. After publishing it, the production query URL should be used.

- Restrict your API keys to specific domains and allow only certain Indexers to query with them

> In [Subgraph Studio](https://thegraph.com/studio/), you can view the following:

Reviewer: ```suggestion
You can also view the following:
```

> - Your user account controls

Reviewer: What does this mean exactly?

- A list of subgraphs you've created
- A section to manage your API keys
- A section to manage, view details, and visualize the status of a specific subgraph

> - A section to manage your billing

Reviewer: If the point of this secondary list is to list what users can _view_ instead of _do_, I find it a bit confusing that most points are about "managing" things. Thinking more about it, I think having a single list would be better, and some of the bullet points could be about viewing information, while others could be about taking actions.

> ## Install the Graph CLI

Reviewer: ```suggestion
## Install The Graph CLI
```

Before deploying, you must install The Graph CLI.

> You must have [Node.js](https://nodejs.org/) and a package manager of your choice (`npm`, `yarn` or `pnpm`) installed to use the Graph CLI. Check for the [most recent](https://github.com/graphprotocol/graph-tooling/releases?q=%40graphprotocol%2Fgraph-cli&expanded=true) CLI version.

Reviewer: ```suggestion
You must have [Node.js](https://nodejs.org/) and a package manager of your choice (`npm`, `yarn` or `pnpm`) installed to use The Graph CLI. Check for the [most recent](https://github.com/graphprotocol/graph-tooling/releases?q=%40graphprotocol%2Fgraph-cli&expanded=true) CLI version.
```

- You can deploy a new version to Studio using the CLI (it will only be private at this point).
- Once you're happy with it, you can publish your new deployment to [Graph Explorer](https://thegraph.com/explorer).
- This action will create a new version of your subgraph that Curators can start signaling on and Indexers can index.

You can also update your subgraph's metadata without publishing a new version. You can update your subgraph details in Studio (under the profile picture, name, description, etc.) by checking an option called **Update Details** in Graph Explorer. If this is checked, an on-chain transaction will be generated that updates subgraph details in Explorer without having to publish a new version with a new deployment.

> Note: There are costs associated with publishing a new version of a subgraph to the network. In addition to the transaction fees, you must also fund a part of the curation tax on the auto-migrating signal. You cannot publish a new version of your subgraph if Curators have not signaled on it. For more information, please read more [here](/network/curating/).

## Automatic Archiving of Subgraph Versions

> Whenever you deploy a new subgraph version in Subgraph Studio, the previous version will be archived. Archived versions won't be indexed/synced and therefore cannot be queried. You can unarchive an archived version of your subgraph in Studio UI.

Reviewer: ```suggestion
Whenever you deploy a new subgraph version in Subgraph Studio, the previous version will be archived. Archived versions won't be indexed/synced and therefore cannot be queried. You can unarchive an archived version of your subgraph in Subgraph Studio.
```

npm install -g @graphprotocol/graph-cli
```

## Create your Subgraph

> ### Create an Account

Reviewer: This heading doesn't seem super relevant; I would expect the steps below to go through the onboarding process including the email verification step, but it only mentions the first step which is to connect the wallet. Furthermore, the sub-headings ("How to Create a Subgraph in Subgraph Studio" and "Subgraph Compatibility with The Graph Network") are not about creating an account but rather about creating a subgraph, so I'm not sure if you meant to make them sub-headings or if they should be at the same level as this one (`###`)? Even Marcus's video doesn't go through the actual "create an account" process since he logs in to an existing account with multiple subgraphs.

> Before deploying your subgraph you need to create a subgraph in [Subgraph Studio](https://thegraph.com/studio/).

Reviewer: Did you mean
```suggestion
Before deploying your subgraph you need to create an account in [Subgraph Studio](https://thegraph.com/studio/).
```
? That would match the heading, but maybe you really did mean "create a subgraph", in which case I would rewrite as:

```suggestion
The first step is to create a subgraph in [Subgraph Studio](https://thegraph.com/studio/).
```

1. Open [Subgraph Studio](https://thegraph.com/studio/).
2. Connect your wallet to sign in.

>    - You can do this via MetaMask, Coinbase Wallet, WalletConnect, or Safe.

Reviewer: ```suggestion
   - You can do this via MetaMask or any other compatible browser extension wallet, Coinbase Wallet, WalletConnect, or Safe.
```

3. After you sign in, your unique deploy key will be displayed on your account home page.

>    - The deploy key allows you to publish your subgraphs or manage your API keys and billing. It is unique but can be regenerated if you think it has been compromised.

Reviewer: That doesn't seem right. The deploy key is shown on subgraph details pages, not on the "account home page", whatever that is (the Settings page?). It _would be nice_ if it was shown on the Settings page, since it is account-specific and not subgraph-specific. 🤔 cc @IainM22

> > Important: You need API keys to query subgraphs

Reviewer: ```suggestion
> Important: You need an API key to query subgraphs.
```

#### How to Create a Subgraph in Subgraph Studio

<VideoEmbed youtube="nGIFuC69bSA" />

> > For additional written detail, review the [quick-start](/quick-start/)

Reviewer: ```suggestion
> For additional written detail, review the [Quick Start](/quick-start/).
```

#### Subgraph Compatibility with The Graph Network

In order to be supported by Indexers on The Graph Network, subgraphs must:

- Index a [supported network](/developing/supported-networks)
- Must not use any of the following features:
  - ipfs.cat & ipfs.map
  - Non-fatal errors
  - Grafting

## Initialize your Subgraph

> Once your subgraph has been created in Subgraph Studio, you can initialize the subgraph code through the CLI using this command:

Reviewer: ```suggestion
Once your subgraph has been created in Subgraph Studio, you can initialize its code through the CLI using this command:
```

```bash
graph init --studio <SUBGRAPH_SLUG>
```

> You can find the `<SUBGRAPH_SLUG>` value on your subgraph details page in Subgraph Studio, see image below:

Reviewer: ```suggestion
You can find the `<SUBGRAPH_SLUG>` value on your subgraph details page in Subgraph Studio, as shown in the following screenshot:
```

![Subgraph Studio - Slug](/img/doc-subgraph-slug.png)

After running `graph init`, you will be asked to input the contract address, network, and an ABI that you want to query. This will generate a new folder on your local machine with some basic code to start working on your subgraph. You can then finalize your subgraph to make sure it works as expected.

## Graph Auth

> Before you can deploy your subgraph to Subgraph Studio, you need to login into your account within the CLI. To do this, you will need your deploy key, which you can find on your "My Subgraphs" page or under your subgraph details page.

Reviewer: ```suggestion
Before you can deploy your subgraph to Subgraph Studio, you need to login into your account within the CLI. To do this, you will need your deploy key, which you can find under your subgraph details page.
```

Then, use the following command to authenticate from the CLI:

```bash
graph auth --studio <DEPLOY KEY>
```

## Deploying a Subgraph

Once you are ready, you can deploy your subgraph to Subgraph Studio.

> > Deploying a subgraph with the CLI pushes it to the Studio, where you can test it and and update the metadata. This action won't publish your subgraph to the decentralized network.

Reviewer: ```suggestion
> Deploying a subgraph with the CLI pushes it to Studio, where you can test it and and update the metadata. This action won't publish your subgraph to the decentralized network.
```

Use the following CLI command to deploy your subgraph:

```bash
graph deploy --studio <SUBGRAPH_SLUG>
```

After running this command, the CLI will ask for a version label.

> - It's strongly recommended to use [semver](https://semver.org/) for versioning like `0.0.1`. That said, you are free to choose any string as version such as:`v1`, `version1`, `asdf`.

Reviewer: ```suggestion
- It's strongly recommended to use [semver](https://semver.org/) for versioning, e.g. `0.0.1`. That said, you are free to choose any string such as `v1`, `version1`, or `asdf`.
```

- The labels you create will be visible in Graph Explorer and can be used by curators to decide if they want to signal on a specific version or not, so choose them wisely.

## Testing your Subgraph

> After deploying, you can test/stage your subgraph in Subgraph Studio, deploy another version, update the metadata, and publish to [Graph Explorer](https://thegraph.com/explorer) when you are ready.

Reviewer: ```suggestion
After deploying, you can test your subgraph (either in Studio, with the playground, or in your own app, with the development query URL), deploy another version, update the metadata, and publish to [Graph Explorer](https://thegraph.com/explorer) when you are ready.
```

> Use Subgraph Studio’s playground to check your subgraph logs and see where a subgraph fails.

Reviewer: Oh no, the playground and the logs are two different things.

## Publish your Subgraph

In order to publish your subgraph successfully, review [publishing a subgraph](/publishing/publishing-a-subgraph/).

## Versioning your Subgraph with the CLI

If you want to update your subgraph, you can do the following:

> - You can deploy a new version to the Studio using the CLI (it will only be private at this point).

Reviewer: ```suggestion
- You can deploy a new version to Studio using the CLI (it will only be private at this point).
```

- Once you're happy with it, you can publish your new deployment to [Graph Explorer](https://thegraph.com/explorer).

> - This action will create a new version of your subgraph that curators can start signaling on and Indexers can index.

Reviewer 2: ```suggestion
- This action will create a new version of your subgraph that Curators can start signaling on and Indexers can index.
```

> You can also update your subgraphs' metadata without publishing a new version. You can update your subgraph details in the Studio (under the profile picture, name, description, etc.) by checking an option called **Update Details** in Graph Explorer. If this is checked, an on-chain transaction will be generated that updates subgraph details in the Explorer without having to publish a new version with a new deployment.

Reviewer: ```suggestion
You can also update your subgraph's metadata without publishing a new version. You can update your subgraph details in Studio (under the profile picture, name, description, etc.) by checking an option called **Update Details** in Graph Explorer. If this is checked, an on-chain transaction will be generated that updates subgraph details in Explorer without having to publish a new version with a new deployment.
```

Also, I find this whole paragraph a bit confusing. Is "Update Details" a checkbox that exists in Studio? I think a screenshot would help greatly here.

> > Note: There are costs associated with publishing a new version of a subgraph to the network. In addition to the transaction fees, you must also fund a part of the curation tax on the auto-migrating signal. You cannot publish a new version of your subgraph if curators have not signaled on it. For more information on the risks of curation, please read more [here](/network/curating/).

Reviewer: ```suggestion
> Note: There are costs associated with publishing a new version of a subgraph to the network. In addition to the transaction fees, you must also fund a part of the curation tax on the auto-migrating signal. You cannot publish a new version of your subgraph if Curators have not signaled on it. For more information on the risks of curation, please read more [here](/network/curating/).
```

> > Note: There are costs associated with publishing a new version of a subgraph to the network. In addition to the transaction fees, you must also fund a part of the curation tax on the auto-migrating signal. You cannot publish a new version of your subgraph if Curators have not signaled on it. For more information on the risks of curation, please read more [here](/network/curating/).

Reviewer 2: ```suggestion
> Note: There are costs associated with publishing a new version of a subgraph to the network. In addition to the transaction fees, you must also fund a part of the curation tax on the auto-migrating signal. You cannot publish a new version of your subgraph if Curators have not signaled on it. For more information on curation, please read more [here](/network/curating/).
```

## Automatic Archiving of Subgraph Versions

Whenever you deploy a new subgraph version in Subgraph Studio, the previous version will be archived. Archived versions won't be indexed/synced and therefore cannot be queried. You can unarchive an archived version of your subgraph in the Studio UI.

> > Note: Previous versions of non-published subgraphs deployed to the Studio will be automatically archived.

Reviewer: ```suggestion
> Note: Previous versions of non-published subgraphs deployed to Studio will be automatically archived.
```

Also, when? As soon as a new version is published, or after some time?
